File: app/backend/src/schemas/role_request.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel

from app.backend.src.enums.role_request_status import RequestStatus
from app.backend.src.enums.user_role import UserRole


class RoleRequestCreate(BaseModel):
    requested_role: UserRole

class RoleRequestReview(BaseModel):
    rejection_reason: Optional[str] = None

class UserSummary(BaseModel):
    id: str
    name: str
    surname: str
    email: str

    class Config:
        from_attributes = True

class ReviewerSummary(BaseModel):
    id: str
    name: str
    surname: str
    
    class Config:
        from_attributes = True


class RoleRequestResponse(BaseModel):
    request_id: str
    user: UserSummary
    requested_role: UserRole
    current_role: UserRole
    status: RequestStatus
    created_at: datetime
    reviewed_at: Optional[datetime] = None
    reviewer: Optional[ReviewerSummary] = None

    class Config:
        from_attributes = True


class RoleRequestList(BaseModel):
    data: List[RoleRequestResponse]
    total: int

File: app/frontend/src/types/role-request.ts
import type { UserRole } from './user';
export type RoleStatus = 'pending' | 'approved' | 'rejected' | 'revoked' | 'none';

export interface UserSummary{
    id: string;
    name: string;
    surname: string;
    email: string;
    license_number: string | null;
}

export interface RoleRequestCreate {
    current_role: UserRole;
}

export interface RoleRequest {
    request_id: string;
    user: UserSummary;
    requested_role: string;
    current_role: string;
    status: RoleStatus;
    firefighter_license_id: string | null;
    created_at: string;
    reviewed_by: string | null;
    reviewed_at: string | null;
}

export interface RoleRequestList {
    data: RoleRequest[];
    total: number;
}